# Lab 02 - Detection, scores and alerts

下午第三節，70 分鐘。這份 notebook 算數字、推資料，並且在每一段之後告訴你去 Grafana 的哪一張
panel 或哪一個編輯畫面上看。主場在 Grafana。

Dashboard: <http://localhost:3000/d/aiops-lab02>，標題 `Lab 02 - Detection, scores and alerts`

**先按 Run All 從頭跑到底，再回到這裡逐段讀。** 整份幾秒鐘就跑完，沒有圖，只有表。publish 那一格
會印出 `published 43,200 rows as 'lab02_detection'`，最後一格把 16 個真值窗當成 annotation 推到
dashboard 上。資料上線之前，Grafana 上的 panel 是空的，後面每一段的判讀都會落空。

早上把 anomaly 定義成「相對於某個明確 baseline 的偏離」，並且把 outlier、anomaly、alert、fault
分開。下午第一節做特徵，第二節做 baseline。這一節把最後一段接起來：score 怎麼變成 label，label
怎麼變成一則會吵醒人的通知，以及這整條鏈路該用哪幾個數字來評估。

三層的分界是整個下午的骨架。score 是連續值，量的是離 baseline 多遠；label 是套上門檻並且連續
確認之後的二元判斷；alert 是通過 duration、最小流量閘門、維護窗、嚴重度與去重之後，真正送到人
手上的那一則。把三者混成一件事，就是告警風暴的來源。

離開這一節的時候，你要能做到三件事。在 Grafana 的告警編輯畫面上指出哪一段是 score、哪一段是
label、哪一段是 alert。自己建一條規則並且說得出 `for` 為什麼設那個值。用 event recall、detection
delay、alerts per day 三個數字替你選的操作點辯護。

| 段落 | 內容 | 分鐘 |
| --- | --- | --- |
| 0 | 環境檢查與資料 | 4 |
| 1 | Layer 1，五種分數與兩把尺 | 7 |
| 2 | 有些特徵不能算 z-score，以及漏掉的那一個事件 | 8 |
| 3 | Layer 2 與 Layer 3，從 dashboard 讀出來 | 5 |
| 4 | 記分卡 | 5 |
| 5 | 政策旋鈕與操作點 | 3 |
| 6 | Grafana Alerting，讀一條現成的規則 | 10 |
| 7 | 自己建第四條規則 | 12 |
| 8 | 維護窗與壓制 | 5 |
| 9 | 嚴重度 | 4 |
| 10 | 一則通知該帶什麼 | 3 |
| 11 | 人的檢查點 | 2 |
| 12 | 沒有真值的那一天 | 2 |


In [1]:
import sys, pathlib

PROJECT_ROOT = pathlib.Path.cwd()
while PROJECT_ROOT != PROJECT_ROOT.parent and not (PROJECT_ROOT / "environment.yml").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent
sys.path.insert(0, str(PROJECT_ROOT / "infra"))

import numpy as np
import pandas as pd
import aiopskit as wk

pd.set_option("display.width", 150)
pd.set_option("display.max_columns", 40)

wk.bootstrap()

project root  /Users/kevin/Library/CloudStorage/GoogleDrive-kevin880987@gmail.com/我的雲端硬碟/workshop/workbench/project/202608-aia-network-monitoring-aiops/aiops-anomaly-zero-to-hero
aiopskit      1.0.0

component             status    detail
------------------------------------------------------------------------------
Prometheus            up        http://localhost:9090
results exporter      up        http://localhost:8010/metrics
Grafana               up        http://localhost:3000  datasource=prometheus
  scrape prometheus     up        
  scrape rrd-exporter   down      
  scrape node-exporter  up        
  aiops_python_result             0 series in Prometheus

Prometheus is running but has no 'python-results-exporter' scrape job, so nothing you
  publish can reach Grafana. This happens when Prometheus was started from a
  package default config instead of the repository one. Stop it and restart with:
    prometheus --config.file=infra/prometheus/prometheus.macos.yml
  (use prometh

PosixPath('/Users/kevin/Library/CloudStorage/GoogleDrive-kevin880987@gmail.com/我的雲端硬碟/workshop/workbench/project/202608-aia-network-monitoring-aiops/aiops-anomaly-zero-to-hero')

## 0. 環境檢查與資料

上一格的 `wk.bootstrap()` 印出四個元件的狀態表，下面接一段修復指示。dashboard 的 panel 空白的話，
依序檢查三件事。

第一，那段指示如果說 `Prometheus is running but has no 'python-results-exporter' scrape job`，
你是用套件預設的設定檔起了 Prometheus，正確的起法在 Lab 00 第 1 節。

第二，exporter 有沒有在跑，`REPLAY_SPEED_X=720 python infra/python_results_exporter.py`。

第三，dashboard 的時間範圍是不是 `now-30m`。所有 panel 都用相對時間，不要去改成絕對區間。

左上角的 `Port` 下拉預設是 `All`，五個 port 的線會疊在一起，先挑一個看比較清楚。

In [2]:
tel, contract = wk.load_telemetry()
print(contract.describe())

CADENCE = contract.cadence_s
W = int(round(3600 / CADENCE))
SPAN = (tel["timestamp"].min(), tel["timestamp"].max())

events = wk.load_events(tel)
calendar = wk.load_change_calendar()
incidents, planned = wk.split_events(events, calendar)

print()
print(f"rolling window  W = {W} samples = {W * CADENCE / 60:.0f} min")
print(f"event windows   {len(events)}  ->  incidents {len(incidents)} (must catch), "
      f"planned {len(planned)} (must stay quiet)")
print()
print(planned[["event_id", "event_type", "port_id", "start", "end", "change_id"]].to_string(index=False))

rows            43,200 across 5 ports
span            2026-02-01 00:00:00  ->  2026-03-02 23:55:00
cadence         300 s per sample (5 min; 288 samples/day)
column reading  delta
monotonic frac  0.503 of consecutive octet steps are non-negative
note            reading chosen automatically: 50.3% of octet steps are non-negative, so the column is a per-interval total, not a counter

rolling window  W = 12 samples = 60 min
event windows   18  ->  incidents 16 (must catch), planned 2 (must stay quiet)



event_id              event_type     port_id               start                 end change_id
       A business_traffic_growth port-id7430 2026-02-04 09:00:00 2026-02-04 10:55:00     CH-02
       C       large_file_backup port-id7431 2026-02-08 01:00:00 2026-02-08 03:55:00     CH-01


### 資料、真值，以及這一節要剖面的對象

`load_telemetry()` 會自己判斷 INOCTETS 這類欄位是累積 counter 還是每個區間的 delta，並印出它採用
了哪一種讀法。不要對這些欄位再做 `.diff()`。

`split_events()` 把 18 個 (event, port) 真值窗切成兩堆。判準在 change calendar 而不在 telemetry：
有人事先申報過的是 planned change，偵測器對它保持安靜才是對的；沒申報的是 incident，必須抓到。
兩者在流量上長得一模一樣。

這一節動手的對象是 score 欄位，所以下一格先把五種分數全部算完，再剖面這些分數本身：值域、來源
特徵的 degenerate fraction、各自越過門檻的頻率，以及正樣本佔比。分數表前面先印這一節 publish 出去
的八個原始特徵，只有空值、零值比例與值域；Lab 01 已經把這些欄位剖得更深，這裡是出貨清單。

同一段流量，五種比法。rolling z 與 robust z 問「跟自己最近一小時比」，seasonal z 問「跟同一個時段
的自己比」，peer z 問「跟兄弟 port 同一瞬間比」，CUSUM 問「這個偏移累積夠久了沒」。它們互相不同意
的地方，就是 baseline 選擇真正在做事的地方。

`score_max` 是後面 label 要用的那一條，它取四個特徵上的五條分數：traffic_bps 的 robust z 與
seasonal z，加上 broadcast_ratio、multicast_ratio、avg_pkt_bytes 各自的 robust z。流量本身看不見
由小封包組成的掃描，也看不見 L2 loop，所以形狀特徵必須進到同一條分數裡。peer z 與 CUSUM 畫在
dashboard 上是給人讀的，不進政策。

peer baseline 要等所有 port 都算完才算得出來，因為它需要同一個 timestamp 上其他四個 port 的值。

In [3]:
# The line the dashboard draws, and the threshold Layer 2 will apply.
# Section 5 sweeps it. This is the one place to change it.
THRESHOLD = 4.0
CUSUM_LIMIT = 1.0

frames = []
for pid, grp in tel.groupby("port_id", observed=True):
    g = grp.sort_values("timestamp").reset_index(drop=True)

    rc, rs = wk.baselines.rolling_mean(g["traffic_bps"], W)
    g["roll_center"], g["roll_lo"], g["roll_hi"] = rc, rc - 3 * rs, rc + 3 * rs
    bc, bs = wk.baselines.rolling_robust(g["traffic_bps"], W)
    g["robust_center"], g["robust_lo"], g["robust_hi"] = bc, bc - 3 * bs, bc + 3 * bs
    sc, ss = wk.baselines.seasonal(g, "traffic_bps", by="daytype_hour")
    g["seasonal_center"], g["seasonal_lo"], g["seasonal_hi"] = sc, sc - 3 * ss, sc + 3 * ss

    g["score_rolling"] = wk.baselines.deviation_score(g["traffic_bps"], rc, rs)
    g["score_robust"] = wk.baselines.deviation_score(g["traffic_bps"], bc, bs)
    g["score_seasonal"] = wk.baselines.deviation_score(g["traffic_bps"], sc, ss)
    g["score_broadcast"] = wk.detect.score_robust_z(g["broadcast_ratio"], W)
    g["score_multicast"] = wk.detect.score_robust_z(g["multicast_ratio"], W)
    g["score_pktsize"] = wk.detect.score_robust_z(g["avg_pkt_bytes"], W)
    g = g.join(wk.detect.score_cusum(g["traffic_bps"], W))
    frames.append(g)

allp = pd.concat(frames, ignore_index=True)

pc, ps = wk.baselines.peer(allp, "traffic_bps")
allp["peer_center"], allp["peer_lo"], allp["peer_hi"] = pc, pc - 3 * ps, pc + 3 * ps
allp["score_peer"] = wk.baselines.deviation_score(allp["traffic_bps"], pc, ps)
# The dashboard queries score_cusum; rename once rather than ship two identical columns.
allp = allp.rename(columns={"cusum_score": "score_cusum"})

SCORES = ["score_robust", "score_seasonal", "score_broadcast", "score_multicast", "score_pktsize"]
allp["score_max"] = allp[SCORES].abs().max(axis=1)
allp["is_incident"] = wk.truth_mask(allp, incidents).astype(float)
allp["is_planned"] = wk.truth_mask(allp, planned).astype(float)

# ---- profile of what this lab will detect on --------------------------------
SCORE_SOURCE = {
    "score_robust": "traffic_bps", "score_seasonal": "traffic_bps",
    "score_broadcast": "broadcast_ratio", "score_multicast": "multicast_ratio",
    "score_pktsize": "avg_pkt_bytes", "score_rolling": "traffic_bps",
    "score_peer": "traffic_bps", "score_cusum": "traffic_bps",
}

degen = {}
for feat in dict.fromkeys(SCORE_SOURCE.values()):
    per_port = []
    for _, g in allp.groupby("port_id", observed=True):
        values = g[feat].reset_index(drop=True)
        _, scale = wk.baselines.rolling_robust(values, W)
        per_port.append(wk.baselines.degenerate_fraction(scale, values))
    degen[feat] = float(np.mean(per_port))

inv = (allp.groupby(["device_id", "port_id", "port_role"], observed=True)
           .agg(samples=("timestamp", "size"), traffic=("traffic_bps", "median"),
                packets_pps_median=("packets_pps", "median")).reset_index())
inv["traffic_kBs_median"] = (inv.pop("traffic") / 1000).round(1)
inv["packets_pps_median"] = inv["packets_pps_median"].round(1)
print("port inventory")
print(inv.to_string(index=False))

steps = (allp.sort_values(["port_id", "timestamp"])
             .groupby("port_id", observed=True)["timestamp"].diff().dt.total_seconds().dropna())
print()
print("coverage and class balance")
print(f"  samples                      {len(allp):>8,}")
print(f"  sampling steps other than {CADENCE:.0f}s  {int((steps != CADENCE).sum()):>3}")
print(f"  inside an incident window    {int(allp['is_incident'].sum()):>8,}   "
      f"{allp['is_incident'].mean():.2%} of samples")
print(f"  inside a planned change      {int(allp['is_planned'].sum()):>8,}   "
      f"{allp['is_planned'].mean():.2%} of samples")

ev_inv = events.copy()
ev_inv["planned"] = ev_inv["event_id"].isin(set(planned["event_id"]))
ev_inv = (ev_inv.groupby(["event_id", "event_type", "planned"], observed=True)
                .agg(ports=("port_id", "nunique"), start=("start", "min"),
                     end=("end", "max"), points=("n_points", "sum")).reset_index())
ev_inv["minutes"] = (ev_inv["points"] / ev_inv["ports"] * CADENCE / 60).round(0).astype(int)
print()
print("event inventory")
print(ev_inv.to_string(index=False))

PUBLISHED = ["traffic_bps", "packets_pps", "error_rate", "discard_rate", "errors_pps",
             "avg_pkt_bytes", "broadcast_ratio", "multicast_ratio"]
raw = []
for feat in PUBLISHED:
    col = allp[feat].astype(float)
    # Ratios and rates need more decimals than bytes per second to stay readable.
    d = 1 if col.max() >= 100 else (3 if col.max() >= 1 else 5)
    raw.append({
        "feature": feat,
        "nulls": int(col.isna().sum()),
        "zero_frac": f"{float((col == 0).mean()):.2%}",
        "min": f"{col.min():,.{d}f}",
        "median": f"{col.median():,.{d}f}",
        "p95": f"{col.quantile(0.95):,.{d}f}",
        "max": f"{col.max():,.{d}f}",
    })
print()
print("raw features this lab publishes, five ports pooled")
print(pd.DataFrame(raw).to_string(index=False))

rows = []
for col, feat in SCORE_SOURCE.items():
    limit = CUSUM_LIMIT if col == "score_cusum" else THRESHOLD
    s = allp[col].abs()
    rows.append({
        "score": col, "source": feat, "in_score_max": col in SCORES,
        "nulls": int(allp[col].isna().sum()),
        "min": round(float(s.min()), 2),
        "median": round(float(s.median()), 2),
        "mean": round(float(s.mean()), 2),
        "p95": round(float(s.quantile(0.95)), 2),
        "max": round(float(s.max()), 1),
        "at_cap_50": f"{float((s >= 49.99).mean()):.2%}",
        "limit": limit,
        "over_limit": f"{float((s > limit).mean()):.2%}",
        "degen_frac": round(degen[feat], 3),
    })
print()
print("score columns, five ports pooled")
print(pd.DataFrame(rows).to_string(index=False))
print()
print(f"score_max over {THRESHOLD}   {float((allp['score_max'] > THRESHOLD).mean()):.2%} of samples")
print(f"cusum signals      {int(allp['cusum_signal'].sum()):,} over the whole month")

port inventory
 device_id     port_id      port_role  samples  packets_pps_median  traffic_kBs_median
core-sw-01 port-id7429  server-uplink     8640               138.8               145.1
core-sw-01 port-id7430    office-vlan     8640                56.3                42.5
dist-sw-02 port-id7431 backup-storage     8640                65.8                76.4
edge-fw-01 port-id7427    wan-primary     8640               111.5                97.9
edge-fw-01 port-id7428  wan-secondary     8640                34.2                27.7

coverage and class balance
  samples                        43,200
  sampling steps other than 300s    0
  inside an incident window         226   0.52% of samples
  inside a planned change            60   0.14% of samples

event inventory
event_id                event_type  planned  ports               start                 end  points  minutes
       A   business_traffic_growth     True      1 2026-02-04 09:00:00 2026-02-04 10:55:00      24      120
      

## 1. Layer 1，五種分數與兩把尺

### 讀這張剖面

`in_score_max` 那一欄把政策的組成攤開：五條分數進 `score_max`，peer z 與 CUSUM 沒進。
`nulls` 數的是滾動窗還沒填滿的樣本，每個 port 的序列前緣都有幾筆，那段時間不可能產生 label。

`at_cap_50` 不是 0.00% 的那幾條分數，已經頂到了 `deviation_score` 的上限 50。頂到上限的讀數要當成
「超出可測量範圍」，不是一個測量值。這件事在第 9 節會回來咬人：嚴重度模型拿 peak score 當輸入，
而 peak score 早就飽和了。

`degen_frac` 是分母的健康檢查，值越高表示尺度估計越常塌到底線，那條分數就越接近在量底線常數，
而不是在量偏離。第 2 節會用單一 port 的完整版本把這件事說完。

`score_rolling` 那一列整個月沒有一筆越過 4.0，最大值只有 3.2，而同一份流量的 robust z 最大到 40
以上。差別在分母：std 被它該抓的那些尖峰自己撐大，MAD 不會。這是 Lab 01 那組 baseline 比較在
分數層級上的樣子。

class balance 那一段是後面所有評估的前提。未申報的 incident 樣本佔全部樣本的 0.52%，正負樣本差
兩個數量級，任何以 point accuracy 為主的評估都會被多數類吃掉。第 4 節會把這個數字換算成一個具體
後果。

### 在 dashboard 上讀 Layer 1

panel `Layer 1 - anomaly score by detector` 上有四條線疊在同一段流量上：rolling z、robust z、
seasonal z、peer z。橫著那條門檻線 4.0 是這四條共用的。

用左上角的 `Port` 下拉切幾個 port，看這四條線在哪裡彼此不同意。它們共用同一份 telemetry，差別只在
跟什麼比。分歧的地方就是 baseline 選擇在做事的地方，這是 Lab 01 那六張 panel 的直接延伸。

### Layer 1b，CUSUM 拿的是另一把尺

再往下一張 `Layer 1b - CUSUM, an accumulating statistic` 之所以獨立，不是排版問題。CUSUM 的決策
界線是 1.0，畫進上面那張圖會讓人拿 4.0 去讀它，而 4.0 對 CUSUM 沒有任何意義。剖面表上 score_cusum
那一列的 `limit` 欄寫的就是 1.0，越線比例與整月訊號次數印在同一格。

差別在記憶。z-score 每來一筆就忘掉上一筆，CUSUM 把標準化殘差一路累加，所以半個 sigma 的持續偏移
單看任何一筆都不像話，累積夠久照樣越線；越線之後那一臂歸零，統計量才是偵測器而不是總帳。

下一格拿事件 A business_traffic_growth 當對照：port-id7430，09:00 開始的兩小時業務量成長。那是
一段真實的成長，形狀是慢慢抬上去然後停在那裡，也是 rolling baseline 最容易吃掉的形狀。讀表的時候
看三欄：traffic_kBs 停在高點沒下來，baseline_kBs 一路追上去，robust_z 於是自己掉回門檻底下。

In [4]:
ramp = allp[(allp["port_id"] == "port-id7430")
            & (allp["timestamp"] >= "2026-02-04 08:50")
            & (allp["timestamp"] <= "2026-02-04 09:45")]

print("event A business_traffic_growth on port-id7430, the ramp starts at 09:00")
print(pd.DataFrame({
    "time": ramp["timestamp"].dt.strftime("%H:%M"),
    "traffic_kBs": (ramp["traffic_bps"] / 1000).round(0),
    "baseline_kBs": (ramp["robust_center"] / 1000).round(0),
    "robust_z": ramp["score_robust"].round(2),
    "cusum": ramp["score_cusum"].round(2),
    "cusum_signal": ramp["cusum_signal"],
}).to_string(index=False))

over_z = ramp.loc[ramp["score_robust"].abs() > THRESHOLD, "timestamp"].dt.strftime("%H:%M").tolist()
over_c = ramp.loc[ramp["cusum_signal"], "timestamp"].dt.strftime("%H:%M").tolist()
print()
print(f"robust z over its {THRESHOLD}    {over_z}")
print(f"cusum over its {CUSUM_LIMIT}       {over_c}")

event A business_traffic_growth on port-id7430, the ramp starts at 09:00
 time  traffic_kBs  baseline_kBs  robust_z  cusum  cusum_signal
08:50         66.0          63.0      0.20   0.69         False
08:55         70.0          65.0      0.41   0.67         False
09:00        131.0          67.0      4.73   1.52          True
09:05        162.0          67.0      7.00   1.30          True
09:10        130.0          69.0      3.33   0.57         False
09:15        125.0          71.0      2.93   1.05          True
09:20        140.0          73.0      3.37   0.57         False
09:25        171.0         100.0      1.36   0.75         False
09:30        140.0         127.0      0.25   0.69         False
09:35        163.0         130.0      0.51   0.70         False
09:40        187.0         135.0      0.66   0.73         False
09:45        135.0         137.0     -0.03   0.62         False

robust z over its 4.0    ['09:00', '09:05']
cusum over its 1.0       ['09:00', '09:05', '09:15

08:55 流量 70 kB/s，baseline 65 kB/s。09:00 跳到 131 kB/s，robust z 4.73；09:05 到 162 kB/s，
z 衝到 7.00。接下來 z 自己掉下去，09:10 是 3.33，09:15 是 2.93，而流量一直維持在 125 到 140，
是事件前的將近兩倍。z 會掉是因為 baseline 追上來了，`robust_center` 從 67 kB/s 一路走到 09:30 的
127 kB/s。robust z 只在 09:00 與 09:05 兩筆越過 4.0，CUSUM 則在 09:00、09:05、09:15 各觸發一次，
多撐的那一筆 z 只有 2.93，站在門檻底下。

後半段同樣要說完。CUSUM 從 09:20 起也不再觸發，因為它累加的殘差是拿同一個 robust baseline 算的，
baseline 一走，兩者一起變盲。累積統計量買到的是持續性，不是免疫力。要對抗被污染的 baseline 得換
baseline，那是 Lab 01 的題目，換一個偵測器解決不了。

## 2. 有些特徵不能算 z-score，以及漏掉的那一個事件

第 4 節的逐事件偵測表裡只有一列 `detected` 是 False：

```text
event_id         event_type     port_id               start                 end
       E link_quality_issue port-id7429 2026-02-14 10:35:00 2026-02-14 11:10:00
```

五條統計分數沒有一條看得見它，原因在分母。

z-score 是 `(x - center) / scale`。當一個指標在絕大多數時間剛好等於零，每個滾動窗的離散度都是零，
`scale` 這時候直接等於程式碼裡用來防止除以零的那個底線常數，資料本身沒有提供任何離散度。此時
任何非零讀數都會拿到一個很大的分數，而那個分數不代表統計意義，它只反映底線設在哪裡。

下一格印 port-id7429 上每個特徵的 degenerate fraction，也就是尺度估計塌到底線的樣本比例，1.000
表示這個特徵的 z-score 完全沒有資訊。同一格接著印固定規則 `errors_pps > 0.05` 在整份資料上掃到
什麼，兩者對照著讀。

In [5]:
FEATURES = ["traffic_bps", "error_rate", "discard_rate", "broadcast_ratio",
            "multicast_ratio", "avg_pkt_bytes", "unknown_proto_pps"]

# Four of these already carry a robust z on every port from the profile cell.
STORED_Z = {"traffic_bps": "score_robust", "broadcast_ratio": "score_broadcast",
            "multicast_ratio": "score_multicast", "avg_pkt_bytes": "score_pktsize"}

p9 = wk.single_port(tel, "port-id7429")
z9 = allp[allp["port_id"] == "port-id7429"]
rows = []
for feat in FEATURES:
    center, scale = wk.baselines.rolling_robust(p9[feat], W)
    stored = STORED_Z.get(feat)
    z = z9[stored] if stored else wk.baselines.deviation_score(p9[feat], center, scale)
    rows.append({
        "feature": feat,
        "degenerate_fraction": round(wk.baselines.degenerate_fraction(scale, p9[feat]), 3),
        "exactly_zero_fraction": round(float((p9[feat] == 0).mean()), 3),
        "max_abs_z": round(float(z.abs().max()), 1),
    })
print("port-id7429, the port event E lands on")
print(pd.DataFrame(rows).to_string(index=False))

# What the fixed rule sees instead. Same metric, no statistics, one hard number.
hot = allp[allp["errors_pps"] > 0.05]
print()
print(f"samples over the fixed rule errors_pps > 0.05:  {len(hot)} of {len(allp):,}")
print(hot.groupby("event_label", observed=True).size().to_string())
print(f"highest errors_pps outside any labelled event:  "
      f"{allp.loc[allp['event_label'] == 'normal', 'errors_pps'].max():.4f}")

port-id7429, the port event E lands on
          feature  degenerate_fraction  exactly_zero_fraction  max_abs_z
      traffic_bps                0.209                  0.000        9.0
       error_rate                0.998                  0.975       50.0
     discard_rate                0.995                  0.914       50.0
  broadcast_ratio                0.001                  0.000       50.0
  multicast_ratio                0.001                  0.000       50.0
    avg_pkt_bytes                0.001                  0.000       29.9
unknown_proto_pps                1.000                  0.987       50.0

samples over the fixed rule errors_pps > 0.05:  20 of 43,200
event_label
link_quality_issue            8
load_sensitive_link_issue    12
highest errors_pps outside any labelled event:  0.0100


`error_rate` 是 0.998。事件 E 的訊號幾乎完全在 error 上，而 error 正好是這份資料裡最不能做
z-score 的欄位。形狀特徵那三個都是 0.001，所以 broadcast 與 multicast 的 z-score 是可信的，
G 與 H 兩個事件在五個 port 上全部被抓到就是這個原因。

順帶注意 `traffic_bps` 的 0.209。流量本身在夜間也會塌：在 port-id7429 上，02:00 那一小時的
degenerate fraction 是 1.000，該小時的平均流量只有 13:00 的十分之一（32.4 kB/s 對 313 kB/s）。
夜裡流量低而平，MAD 頂在底線上，一點點絕對變化就換算成很多個 sigma。

這裡有一個命名上的坑要記住。欄位叫 `traffic_bps`，但 `load_telemetry()` 是把 INOCTETS 除以
cadence 得到的，單位是每秒 octet 也就是 bytes per second，不是 bits per second。Grafana panel
的單位設成 megabytes per sec 才是對的。看軸的時候不要自己乘八。

### 到 Grafana 上看

切到 Lab 01 dashboard <http://localhost:3000/d/aiops-lab01>，`Port` 選 `port-id7429`，找 panel
`Quality features: error and discard rate`。你會看到一條長時間貼著零的線，偶爾跳起來。這種形狀
沒有可用的統計尺度，它有的是一個明確的工程邊界：介面不該有 error。

回到 Alerting > Alert rules，打開第三條 `Interface errors present`。它的條件是 `errors_pps > 0.05`，
固定門檻，打在原始計數上，完全沒有統計。

### 這條規則的成本效益

整整 30 天、43,200 個樣本裡，`errors_pps` 超過 0.05 的只有 20 筆，全部落在兩個事件上：E 佔 8 筆，
F load_sensitive_link_issue 佔 12 筆。兩段各自連續，所以這條規則一個月只會產生兩則告警。歸類為
normal 的樣本裡，`errors_pps` 的最大值是 0.0100，門檻 0.05 站在健康讀數最高點的五倍之上；事件 E
期間的讀數在每秒 1.7 到 1.9 之間，是健康上限的將近兩百倍。零誤報。

拿它跟另一條路比。想用統計分數抓到 E，門檻要從 4.0 降到 3.0，此時 event recall 補滿到 1.000，
代價是一個月的告警從 182 則變成 511 則，每天從 6.07 則變成 17.03 則。多抓一個事件，多付 329 則
告警。

這就是這個下午最重要的一句話：**一個帶有明確且已知有意義邊界的指標，要的是規則，不是統計。**
把統計門檻往下調，是在用整個團隊的注意力換一個本來就該用別的方法解決的問題。

順便說明，門檻降到 3.0 之所以能抓到 E，靠的是 10:40 那筆 `score_pktsize` 3.36 和 10:45 那筆
`score_robust` 3.58，兩筆分別由不同特徵貢獻。那是巧合，不是偵測。真正的訊號在 error 上，而錯誤率
的量級差異大到不需要任何統計就分得開。

### 那 ML 呢

`score_max` 是四個特徵上五條 z 分數取最大，這種組合的好處是任何一則告警都能立刻回答「是哪一個
特徵、偏離幾個 sigma」，而一個 ML 層買到的是跨特徵的聯合結構，例如 broadcast 微升同時
avg_pkt_bytes 微降這種單看任一條都不越線的組合。代價在凌晨三點才會顯現：你沒辦法對著一個
isolation forest 的輸出說出它為什麼開口，而值班的人需要的正是那句話。今天不上這一層是選擇不是
疏忽，想看它長什麼樣，`labs/self-study/04_ml_anomaly_detection.ipynb` 在那裡。

## 3. Layer 2 與 Layer 3，從 dashboard 讀出來

### Layer 2: label

`confirm()` 回傳兩欄，因為它們回答不同的問題。`breach` 是分數越過門檻的樣本；`label` 是偵測器願意
把話說出口的樣本。兩者的差就是 2 筆連續確認替你擋掉的單點雜訊，代價是最快的偵測也慢了一個取樣
週期，也就是 5 分鐘。

門檻 4.0 定義在剖面那一格，連續 2 筆定義在下一格。兩個都是這份講義的出發點，不是答案。第 5 節的
sweep 會把整條操作曲線攤開來，你要在那裡自己選一個點。

In [6]:
N_CONSEC = 2

flags = wk.confirm(allp["score_max"], THRESHOLD, n_consecutive=N_CONSEC, direction="high")
allp["label"] = flags["label"].astype(float)

n_breach = int(flags["breach"].sum())
n_label = int(flags["label"].sum())
print(f"samples            {len(allp):>7,}")
print(f"breach  (score > {THRESHOLD})  {n_breach:>7,}   {n_breach / len(allp):.3%} of samples")
print(f"label   (x{N_CONSEC} consecutive)  {n_label:>7,}   {n_label / len(allp):.3%} of samples")
print(f"removed by confirmation  {n_breach - n_label:>7,}")

samples             43,200
breach  (score > 4.0)    1,500   3.472% of samples
label   (x2 consecutive)      437   1.012% of samples
removed by confirmation    1,063


### Layer 3: alert

label 到 alert 之間放的全是營運條件。最小流量閘門的用途是保護比率型指標：凌晨四點五個封包裡有三個
是錯的，錯誤率 60%，沒有人需要為此起床。`gap_tolerance` 讓一次事件中間掉一筆不會裂成兩則告警。
change calendar 比對之後被壓下來的候選告警仍然留在表裡，帶著 `suppressed_by`，因為看不見的壓制
沒辦法稽核：一條默默吞掉真實事件的規則，跟一條運作正常的規則，從外面看長得一樣。

用途歸用途，這兩個旋鈕在這份資料上都沒有真的動到任何一則告警，第 5 節會量給你看。先照原設定跑完，
後面比較才有基準。

嚴重度用 deviation x duration x business impact 算，三個乘數都在 `wk.detect._severity` 裡，
第 9 節會把那條公式和它藏的一個缺陷拆開。

In [7]:
policy = wk.AlertPolicy(min_consecutive=N_CONSEC, gap_tolerance=1,
                        min_volume_col="packets_pps", min_volume=1.0)

alerts = wk.build_alerts(allp, "label", "score_max", policy=policy, cadence_s=CADENCE,
                         changes=calendar, method="multi_feature",
                         evidence_cols=tuple(SCORES))

allp["alert"] = 0.0
for a in alerts[alerts["notified"]].itertuples():
    hit = ((allp["port_id"] == a.port_id) & (allp["timestamp"] >= a.fire_time)
           & (allp["timestamp"] <= a.clear_time))
    allp.loc[hit, "alert"] = 1.0

print(f"alert candidates from label runs  {len(alerts)}")
print(f"suppressed by the change calendar {int((~alerts['notified']).sum())}")
print(f"notified                          {int(alerts['notified'].sum())}")
print()
print(alerts[alerts["suppressed_by"] != ""][
    ["port_id", "port_role", "fire_time", "duration_s", "peak_score", "severity", "suppressed_by"]
].to_string(index=False))

alert candidates from label runs  184
suppressed by the change calendar 2
notified                          182

    port_id      port_role           fire_time  duration_s  peak_score severity suppressed_by
port-id7430    office-vlan 2026-02-04 09:05:00      8400.0   50.000000 critical         CH-02
port-id7431 backup-storage 2026-02-08 01:05:00     12000.0   42.756582 critical         CH-01


### 在 dashboard 上讀這兩排狀態

panel `Layer 2 - label, and Layer 3 - alert after policy` 有兩排狀態。上排是 label，也就是分數
越過 4.0 並且連續兩筆成立的樣本。下排是 alert，也就是通過整套政策之後真正會送出去的區間。

先別往下讀。看著這兩排，當場說出上下之間差掉的那些是被什麼拿掉的。

上面幾格印過的三組數字合起來就是答案。43,200 個樣本裡，分數越過 4.0 的有 1,500 筆，佔 3.47%。
要求連續兩筆之後剩下 437 筆，佔 1.01%，中間被擋掉 1,063 筆。那 1,063 筆全部是單點雜訊，一筆超標、
下一筆就回來，沒有一個真實的網路事件是這種形狀。437 筆再收斂成 184 段連續區間，其中 2 段落在已
申報的變更窗裡被壓下來，剩下 182 則送出。

所以落差只有兩層來源：連續確認拿掉單點抖動，變更行事曆拿掉已經有人申報過的工作。兩者都不是統計，
都是營運決定。`AlertPolicy` 上還掛著第三個閘門，最小流量，出廠設定下它一筆都沒有擋掉，第 5 節會
把它轉到會咬人的位置。

### 對照真值

往下捲到 `Ground truth - known incident windows`，這一排在真值窗裡是 1。把它和上面的 alert 排
一起看：alert 平的、真值高的地方就是漏抓，反過來就是誤報。

最後一格推上去的 annotation 是同一份真值的另一種畫法。它掛在 dashboard 上而不是掛在
某一條 series 上，所以每一張 panel 都會出現同一批灰底區間，`Evidence - traffic against the
baseline that fired` 那張圖上因此可以一眼看出尖峰有沒有落在區間裡。看不到就回去重跑那一格，它算的
是當下這一輪重播的位置，過了一輪要重推一次。

`is_planned` 那條線只在兩個地方亮起來，A business_traffic_growth 與 C large_file_backup。這兩個
事件在流量上和真正的故障長得一模一樣，唯一的差別是有人事先在 `change_calendar.csv` 裡申報過。
偵測器對它們保持安靜才是正確行為。

## 4. 記分卡

下一格印三段。第一段是完整的記分卡，dashboard 上四塊 stat tile 的數字都從這裡來。第二段把
accuracy 和 event recall 並排。第三段是逐事件偵測表。

真正該看的三個數字是 event recall、detection delay 與 alerts per day。事件層級的 recall 問的是
「這 16 個 incident 裡有幾個至少被抓到一次」，而不是「事件的每一個取樣點都被標記了嗎」。一個跑
三十分鐘的事件不需要每一筆都命中，它需要被抓到一次，夠早，而且不要淹掉值班的人。

false alert 分成兩堆也是刻意的。落在已申報變更窗裡而沒被壓制的，是壓制設定的失誤，修法已知；落在
平常流量上的，是模型的失誤。合在一起報，就看不出你手上是哪一種問題。

逐事件偵測表裡的 `detection_delay_min` 是真值窗起點到第一則告警的時間。G broadcast_storm 與
H multicast_flooding 各佔五列，因為它們同時打在五個 port 上。

In [8]:
truth = allp["is_incident"].astype(bool)
summary, ev, al = wk.score(alerts, incidents, SPAN, labels=flags["label"], truth=truth,
                           changes=calendar, method="multi_feature")

print("operational scorecard")
for k in ["events_total", "events_detected", "event_recall",
          "alerts_total", "alerts_per_day", "median_detection_delay_min",
          "max_detection_delay_min", "duplicate_alerts_per_event",
          "alerts_suppressed_by_policy", "false_alerts",
          "false_alerts_on_planned_change", "false_alerts_unexplained"]:
    v = summary[k]
    print(f"  {k:<34} {v:.4f}" if isinstance(v, float) else f"  {k:<34} {v}")

rate = float(allp["is_incident"].mean())
print()
print("why accuracy is the wrong headline")
print(f"  event_recall                       {summary['event_recall']:.4f}")
print(f"  point_accuracy                     {summary['point_accuracy']:.4f}")
print(f"  point_precision                    {summary['point_precision']:.4f}")
print(f"  point_recall                       {summary['point_recall']:.4f}")
print(f"  accuracy of a detector that never fires  {1 - rate:.4f}")

show = ev[["event_id", "event_type", "port_id", "port_role",
           "detected", "n_alerts", "detection_delay_s", "peak_score"]].copy()
show["detection_delay_min"] = show.pop("detection_delay_s") / 60
print()
print("per-event detection")
print(show.sort_values(["event_id", "port_id"]).round(2).to_string(index=False))

print()
print("missed incidents")
print(ev[~ev["detected"]][["event_id", "event_type", "port_id", "start", "end"]].to_string(index=False))

operational scorecard
  events_total                       16
  events_detected                    15
  event_recall                       0.9375
  alerts_total                       182
  alerts_per_day                     6.0674
  median_detection_delay_min         5.0000
  max_detection_delay_min            5.0000
  duplicate_alerts_per_event         1.0667
  alerts_suppressed_by_policy        2
  false_alerts                       166
  false_alerts_on_planned_change     0
  false_alerts_unexplained           166

why accuracy is the wrong headline
  event_recall                       0.9375
  point_accuracy                     0.9891
  point_precision                    0.2174
  point_recall                       0.4204
  accuracy of a detector that never fires  0.9948

per-event detection
event_id                event_type     port_id      port_role  detected  n_alerts  peak_score  detection_delay_min
       B         small_packet_scan port-id7428  wan-secondary      True        

### 讀記分卡

`Event recall` 是 0.9375，tile 的 unit 設成 percentunit，所以畫面上顯示的是 93.75%，16 個必須
抓到的 incident 只漏了一個。一個月送出 182 則，`Alerts per day` 6.07。`Median detection delay`
停在 5 分鐘，剛好一個取樣週期。第四塊 `Unexplained false alerts` 數到 166。

延遲那一格值得停一下。5 分鐘是這份資料的 cadence，代表分數在事件的第一筆就越線，第二筆確認完立刻
送出。這是這個設定下能達到的最快值，快不過取樣本身。要更快只有一條路，把 cadence 縮短，而那是
採集端的成本，不是演算法問題。

誤報那一格更值得停。182 則裡只有 16 則對應到已知事件，event precision 0.088，大約每 11 則告警裡
有 1 則指向真的東西。

第五個數字沒有 tile，只印在記分卡裡：`duplicate_alerts_per_event` 1.07。它衡量的是禮貌，問的是
同一個事件送了幾則通知，理想值 1.00。1.07 已經很接近，因為 15 個被抓到的事件裡只有
B small_packet_scan 裂成兩則。第 5 節的 `gap_tolerance` 表會告訴你把它壓到 1.00 要付多少錢。

### 為什麼 accuracy 是錯的數字

未申報的 incident 樣本佔全部樣本的 0.52%，剖面那一格印過。一個永遠不開口的偵測器，point
accuracy 是 0.9948，而它抓到的事件是零。記分卡把 `point_accuracy` 印在 `event_recall` 旁邊就是
為了這個對照：本次設定的 point_accuracy 是 0.9891，比那個什麼都不做的偵測器還低。如果你的 KPI
是 accuracy，你會被獎勵去關掉偵測器。

point precision 0.217 與 point recall 0.420 也不該當成頭條，但它們至少有診斷價值。point recall
只有 0.42 表示事件期間有一半以上的取樣點沒有進到 label，而 event recall 仍然是 0.9375。兩者不
矛盾，這正是為什麼時間序列監控要用事件層級的指標。

### 出聲回答一次

現在請每個人對著自己的團隊回答，出聲講：

> 一天 6 則告警，我的 on-call 付得起嗎？如果其中十一分之一才是真的呢？

沒有標準答案。一個三人輪班、有值班費、有 runbook 的團隊，一天 6 則是可以吸收的。一個把告警轉進
共用信箱、沒有人專責的團隊，一天 6 則兩週之後就會變成沒有人看。這個判斷決定了你在第 5 節那張
sweep 表上該站哪一格，它是營運現實，不是統計量。

## 5. 政策旋鈕與操作點

### 兩個空轉的旋鈕

上面的記分卡是在 `min_volume=1.0` 與 `gap_tolerance=1` 之下算出來的。3.37 是 `packets_pps` 在
43,200 筆裡的全域最小值，第一百分位 6.17，中位數 64.92，所以那個閘門站在所有樣本的下方，一筆都
擋不到。`gap_tolerance` 從 0 改到 1 也一則告警都不差。把那兩行整段刪掉，記分卡一個數字都不會變。

空轉的旋鈕比設錯的旋鈕更難處理，因為它讓人以為問題已經有人管了。下一格把它們一路往上推，你要在
第一張表上找懸崖：閘門抬到哪一格開始拿漏掉的事件換安靜。

In [9]:
def scorecard(min_volume=1.0, gap_tolerance=1, critical_score=2.0):
    pol = wk.AlertPolicy(min_consecutive=N_CONSEC, gap_tolerance=gap_tolerance,
                         min_volume_col="packets_pps", min_volume=min_volume,
                         critical_score=critical_score)
    made = wk.build_alerts(allp, "label", "score_max", policy=pol, cadence_s=CADENCE,
                           changes=calendar, method="multi_feature")
    got, _, _ = wk.score(made, incidents, SPAN, labels=flags["label"], truth=truth,
                         changes=calendar)
    return made, got


print(f"packets_pps  min {allp['packets_pps'].min():.2f}  "
      f"1st pct {allp['packets_pps'].quantile(0.01):.2f}  "
      f"median {allp['packets_pps'].median():.2f}")
print(f"samples the shipped gate (min_volume={policy.min_volume}) excludes: "
      f"{int((allp['packets_pps'] < policy.min_volume).sum())}")

rows = []
for mv in [1, 10, 20, 50, 100]:
    _, got = scorecard(min_volume=float(mv))
    rows.append({"min_volume": mv, "alerts": got["alerts_total"],
                 "per_day": round(got["alerts_per_day"], 2),
                 "event_recall": round(got["event_recall"], 4),
                 "unexplained_fp": got["false_alerts_unexplained"]})
print()
print("raising the minimum volume gate")
print(pd.DataFrame(rows).to_string(index=False))

rows = []
for gt in [0, 1, 2, 3, 5]:
    _, got = scorecard(gap_tolerance=gt)
    rows.append({"gap_tolerance": gt, "alerts": got["alerts_total"],
                 "duplicates_per_event": round(got["duplicate_alerts_per_event"], 2),
                 "event_recall": round(got["event_recall"], 4)})
print()
print("merging label runs across a gap")
print(pd.DataFrame(rows).to_string(index=False))

packets_pps  min 3.37  1st pct 6.17  median 64.92
samples the shipped gate (min_volume=1.0) excludes: 0



raising the minimum volume gate
 min_volume  alerts  per_day  event_recall  unexplained_fp
          1     182     6.07        0.9375             166
         10     172     5.73        0.9375             156
         20     161     5.37        0.9375             145
         50     129     4.30        0.7500             116
        100      96     3.20        0.6250              85



merging label runs across a gap
 gap_tolerance  alerts  duplicates_per_event  event_recall
             0     182                  1.07        0.9375
             1     182                  1.07        0.9375
             2     178                  1.00        0.9375
             3     174                  1.00        0.9375
             5     171                  1.00        0.9375


懸崖落在 20 和 50 中間。抬到 20，送出去的告警從 182 掉到 161，少了 21 則，而 recall 一動也不動；
以候選區間算，消失的有 25 段，同時裂出 8 段新的，淨少 17 段；消失的那 25 段裡有 23 段起點落在
00:00 到 05:59。這不是巧合，packets_pps 的中位數在 02:00 是 16.1、在 13:00 是 151.7，一個絕對
流量閘門本來就是一把時段刀。抬到 50 才開始要付錢：recall 從 0.9375 掉到 0.7500，掉的是
H multicast_flooding 在三個 port 上的那三段。H 從 20:05 打到 21:30，而 packets_pps 的中位數在
20 點是 64.1、21 點是 44.7，閘門設 50 正好切在事件中間。

要帶走的不是「min_volume 設 20」。20 是這份資料上的答案，換一組 port 就不是。要帶走的是找懸崖的
方法：把閘門一路往上推，盯著 recall 那一欄，它第一次動的前一格就是你的上限。順帶把這件事記清楚，
這個閘門擋不到那 166 則誤報的主體。166 則裡只有 32 則落在 00:00 到 05:59，其中 23 則擠在 00 點
那一小時，其餘一百三十幾則發生在流量充足的白天。白天那些是 degenerate scale 和形狀特徵的問題，
跟分母大小無關。一個保護比率型指標的閘門，在別的資料集上會很有用，在這一份上你得先讓它咬得到。

第二張表回答上一節留下的問題。`gap_tolerance` 從 1 改成 3，`duplicate_alerts_per_event` 降到
1.00，而告警總數從 182 掉到 174。少掉的那 8 則裡，哪幾則是被合併回原本的事件、哪幾則是整段消失，
這是你要自己分清楚的。

### 操作點掃描，這一節的核心練習

門檻是營運問題。往下拉會拿到更高的 event recall 和更短的 detection delay，代價是每天更多的
告警；往上推則反過來。這個交換率才是你真正在設計的東西，門檻只是它被寫下來的形式。

看表的時候問兩個問題。從 4.0 降到 3.0，多抓到幾個事件，多付出幾則告警。你的團隊一天能真正讀完
幾則告警，這個數字決定了你能站在曲線的哪一段。

In [10]:
def make_alerts(frame, label_col, score_col):
    return wk.build_alerts(frame, label_col, score_col, policy=policy,
                           cadence_s=CADENCE, changes=calendar, method="sweep")

curve = wk.sweep(allp, "score_max", incidents, SPAN,
                 thresholds=[2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 6.0, 8.0],
                 build_alerts_fn=make_alerts, truth=truth, n_consecutive=N_CONSEC)

print(curve[["threshold", "events_detected", "event_recall", "alerts_total", "alerts_per_day",
             "median_detection_delay_min", "false_alerts", "point_accuracy"]]
      .round(3).to_string(index=False))

 threshold  events_detected  event_recall  alerts_total  alerts_per_day  median_detection_delay_min  false_alerts  point_accuracy
       2.5               16         1.000          1065          35.504                         5.0          1048           0.958
       3.0               16         1.000           511          17.035                         5.0           494           0.978
       3.5               15         0.938           291           9.701                         5.0           275           0.986
       4.0               15         0.938           182           6.067                         5.0           166           0.989
       4.5               15         0.938           133           4.434                         5.0           118           0.991
       5.0               15         0.938            94           3.134                         5.0            79           0.992
       6.0               15         0.938            58           1.934                   

最後一欄把第 4 節的論證補完。門檻從 2.5 推到 8.0，accuracy 從 0.958 單調爬到 0.995，而 event
recall 從 1.000 掉到 0.9375、alerts per day 從 35.5 掉到 1.2。accuracy 一路上升的那個方向，正好
是偵測能力一路下降的方向。這個指標對你的決策沒有提供任何資訊。

選好你的操作點，回到剖面那一格改 `THRESHOLD`，從那裡重跑一次，然後準備替你選的那個點辯護。
辯護要用第 4 節那三個數字，加上你自己團隊一天能讀完幾則。

## 6. Grafana Alerting，讀一條現成的規則

左側 nav > Alerting > Alert rules，folder `AIOps Workshop`，group `lab02`，打開
`Traffic deviation from robust baseline`。

### 三段對應三層

規則的 query 區有 A、B、C 三段，往下讀會發現它們的邊界跟早上那張投影片畫在同一個地方。

**A** 是 Prometheus query，`aiops_python_result{column="score_robust"}`，相對時間範圍 15 分鐘。
這一段拿到的是 score，一個連續值，每個 port 一條 series。它不做任何判斷。

**B** 是 `Reduce`，reducer 選 `last`。它把 A 的一條時間序列壓成一個數字，取最後一筆。這一段決定
「用哪個時刻的值來判斷」，選 `last` 表示只看最新，選 `max` 就會變成「過去 15 分鐘裡有沒有任何
一刻超標」，那是不同的偵測器。

**C** 是 `Threshold`，`IS ABOVE 4`。到這裡才產生 label，一個二元判斷。

alert 不在 A、B、C 裡。alert 是 C 為真之後，經過 pending period、notification policy、mute
timing、contact point 之後真正送出去的那一則。這正是早上說的「label 不等於 alert」，Grafana 的
介面把這條界線畫得很清楚：條件在 query 區，其餘在下面的 evaluation 與 notification 區。

### `for` 是 deadband，而且是時間不是筆數

往下捲到設定評估行為的那一段，找到控制 `for` 的欄位，Grafana 13 上標成 pending period。現在是
`10m`。

`for` 的意思是：條件必須連續為真達到這個時間長度，規則才從 Pending 轉成 Firing。它是一個
duration，不是「連續 N 筆」。這兩者只有在你知道 cadence 的時候才能互換。這份資料 300 秒一筆，
所以 `10m` 等於 2 筆；如果哪天採集改成 60 秒一筆，同樣的 `10m` 就變成 10 筆，規則的行為在你沒有
動它的情況下改變了。這是設定漂移最常見的來源之一，寫 runbook 的時候要把 cadence 一起寫進去。

### 動手：改 `for`，看狀態機

先講一件會擋住你的事。exporter 以 720 倍速重播，牆上的一分鐘等於資料裡的 12 小時。`for: 10m` 用
牆鐘計算，換算成資料時間是 5 天。任何一個 90 分鐘的事件在牆上只存在 7.5 秒，所以在重播環境裡直接
把 `10m` 改成 `5m`，你什麼都看不到，兩個值都遠大於事件本身。原因是重播的時間壓縮和 pending
period 的時間基準不同，設定本身沒有錯。

要看到狀態機真的動，用一個條件恆真的暫時設定：

1. 在 C 段把門檻從 `4` 改成 `-100`。deviation score 上下限是 ±50，所以這個條件永遠成立。
2. pending period 改成 `1m`，儲存。
3. 回到 Alerting > Alert rules，展開這條規則。狀態會先變 Pending，約一分鐘後轉成 Firing。
   展開之後每個 port 是一個 instance，五條 series 五個狀態。
4. pending period 改成 `5m`，儲存。狀態回到 Normal 再重新走一次，這次要等五分鐘才轉 Firing。
   不必真的等滿，看它停在 Pending 就夠了。
5. **全部改回去**：C 段門檻 `4`，pending period `10m`。

第 3 步和第 4 步之間的差別，就是把 `for` 從 `10m` 改成 `5m` 會發生的事，只是搬到一個在教室裡看得見
的時間尺度上。deadband 越長，Pending 停留越久，能擋掉的抖動越多，你也越晚知道有事發生。這條交換
和門檻那條是同一類，只是換一個維度。

順手確認兩件事。規則 group 上有一個評估間隔，Grafana 預設一分鐘，打開看它顯示什麼，因為 pending
period 實際上是以評估次數為單位在數的，設成小於評估間隔的值沒有意義。另外，這個 metric 帶著
`event_label` 這個 label，那是真值，production 不會有。不要在任何規則裡用到它。

## 7. 自己建第四條規則

接下來換你動手，目標是 multicast flooding，12 分鐘內完成，並且能替每個選擇說出理由。

先在 Explore 把 query 試出來，開法在 Lab 00 第 7 節：

```promql
aiops_python_result{column="score_multicast"}
```

應該回來五條 series，一個 port 一條。確認有資料再去建規則，不要在告警編輯畫面裡除錯 query。

接著 Alerting > Alert rules > New alert rule：

1. 規則名稱寫 `Multicast share elevated`，或任何說得出是什麼的名字。
2. A 段貼上上面那條 PromQL，查詢編輯器切到 `Code` 模式。
3. B 段 `Reduce`，reducer `last`。
4. C 段 `Threshold`，`IS ABOVE`，門檻自己決定，理由見下。
5. folder 選 `AIOps Workshop`，evaluation group 選既有的 `lab02`。
6. pending period 自己決定，理由見下。
7. labels 加 `severity`，值自己決定。annotation 至少寫一句 summary，帶上 `{{ $labels.port_id }}`。
8. 儲存，回列表確認它出現在 `lab02` group 裡。

### 門檻要怎麼選

剖面那一格印過各分數的分布。`score_multicast` 的絕對值中位數是 0.67，平均 0.99，最大值 50，
而 50 是 `deviation_score` 的上限，代表已經超出可測量範圍。

現成的 `Broadcast share elevated` 用 6.0，比 `score_robust` 那條的 4.0 高。理由是形狀比率的
baseline 很窄，degenerate fraction 只有 0.001，正常時候幾乎不動，所以同樣一個絕對變化換算出來的
sigma 數比流量大得多。門檻設高一點，是把兩個尺度不同的分數放回可比較的位置上。你可以選
5、6 或 8，但要說得出你在跟什麼比。

### `for` 要怎麼選，用事件 H 來論證

事件 H multicast_flooding 在 catalog 裡是 18 個 period，一個 period 300 秒，所以它實際持續
90 分鐘，而且同時打在五個 port 上。

上限這樣算。`for` 必須遠小於你打算抓的最短事件，否則規則永遠等不到轉 Firing 就結束了。這份資料裡
最短的事件是 E，8 個 period，40 分鐘。所以任何超過 40 分鐘的 pending period 會讓那一類事件在你的
規則上完全不存在，而且不會有任何錯誤訊息告訴你這件事。即使只針對 H，`for` 設 60 分鐘也只剩
30 分鐘可以反應。

下限這樣算。cadence 是 300 秒，`for` 小於 300 秒等於沒有 deadband，一筆雜訊就送出通知。最小可用值
是 2 筆，也就是 `10m`，這和 `confirm()` 那一格的 `n_consecutive=2` 是同一個決定，只是換一種寫法。

所以合理區間是 10 到 15 分鐘。設 `10m`，H 的 90 分鐘裡花 10 分鐘確認，佔事件壽命的 11%，留
80 分鐘給人反應。設 `15m` 多換一點抗雜訊能力，代價是最短的事件只剩 25 分鐘。兩個選擇都可以辯護，
設 `1h` 不行。

寫完之後跟旁邊的人交換，讓對方問你為什麼是這個數字。答不出來就是還沒設計完。

## 8. 維護窗與壓制

左側 nav > Alerting > Notification policies，切到 `Time intervals` tab，打開
`aiops-maintenance-window`。時間區間是 00:30 到 05:00，沒有指定星期，所以是每天。Grafana 13 在
畫面上把這種物件叫 time interval，但 route 還是 `/alerting/routes/mute-timing/edit`，
provisioning API 也還是 `/api/v1/provisioning/mute-timings`，所以看文件或寫自動化的時候，你要找的
關鍵字仍然是 mute timing。

### 它在保護什麼

`change_calendar.csv` 裡的 CH-01：

```text
CH-01,scheduled_backup,port-id7431,2026-02-08 00:50,2026-02-08 04:10,infra-ops,週日夜間全量備份至異地儲存,sustained traffic surge on backup-storage
```

port-id7431 是 backup-storage。週日凌晨的全量備份會在這個 port 上產生真實而且很大的流量突增，
在事件表裡它是 C large_file_backup，01:00 到 03:55，peak score 42.76，持續 180 分鐘 36 筆。
這是真的流量，偵測器抓到它是正確的；把人叫起來則是錯的。

### 兩層壓制，位置不同

`build_alerts` 的壓制發生在偵測端：拿 change calendar 比對，符合的候選告警被標上
`suppressed_by=CH-01`，`notified=False`，第 3 節那格印出來的就是這兩筆。

Grafana 的 mute timing 發生在通知端。被 mute 的規則照樣評估、照樣進 Firing、照樣顯示在 Alerting
介面上，只是不送出通知。事後檢討的時候這個差別很重要：「規則沒發現」和「規則發現了但我們決定不
吵人」是兩種完全不同的失效，前者要改偵測，後者要改流程。

### 練習：把窗縮成 02:00 到 05:00

真的存檔之前先想清楚，再動手。

CH-01 從 00:50 開始，事件 C 的流量從 01:00 起。壓制窗改成 02:00 之後，01:00 到 02:00 這一小時重新
變成會送出通知的區間。以 300 秒 cadence 算是 12 筆樣本，以 `for: 10m` 計，第一則通知大約在 01:10
送出，而值班的人會在週日凌晨一點看到 backup-storage 流量暴增的告警，去查之後發現是 infra-ops
排定的備份。

這個練習的重點是**壓制窗必須從變更記錄推導，不能憑印象設**。CH-01 申報的起點
是 00:50 而不是 01:00，備份腳本開始跑和流量真的上來之間有十分鐘的差距。壓制窗如果貼著事件實際起點
設，你就會在下一次腳本早跑五分鐘的時候被叫醒。

### 重播環境下的注意事項

mute timing 依真實時鐘判定，重播把資料時間壓縮了 720 倍，資料裡的 00:50 和牆上的 00:50 沒有對應
關係。所以這一段在課堂上是設定與推理，不是即時驗證。想真的看到 mute 生效，把時間區間改成接下來
十分鐘的牆鐘時間，配合第 6 節那個恆真條件，你會看到規則是 Firing 但通知沒有送出。做完記得改回
00:30 到 05:00。

另外，這個 time interval 沒有指定時區。編輯畫面上的欄位是 Time range、Days of the week、
Days of the month、Months、Years 和 Location，IANA 時區填在 Location，空白就是 UTC。打開看它現在
顯示什麼，記下來，因為一個跨時區的團隊在這個欄位上吃過的虧比在門檻上多。

## 9. 嚴重度

早上給的公式是 severity = deviation x duration x business impact。`wk.detect._severity` 把它寫成
可以計算的形式：

```text
severity_score = 0.5 x min(peak_score / critical_score, 2)
               + 0.3 x min(duration_s / critical_duration_s, 2)
               + 0.4 x role_weight

critical: >= 1.0     warning: >= 0.55     info: 其餘
```

下一格印三段：出廠設定下的嚴重度分布、`wk.detect.DEFAULT_ROLE_WEIGHT` 這張權重表，以及
`critical_score` 移動時分布怎麼變。第一段裡藏著一個現成的缺陷，請當場找出來再往下讀。

In [11]:
print("severity as shipped (critical_score=2.0)")
print(alerts["severity"].value_counts().to_string())

print()
print("role weights in wk.detect.DEFAULT_ROLE_WEIGHT")
for role, weight in wk.detect.DEFAULT_ROLE_WEIGHT.items():
    print(f"  {role:<16} {weight}")

print()
print("severity split as critical_score moves")
for cs in [2.0, 8.0, 12.0, 15.0]:
    made, _ = scorecard(critical_score=cs)
    counts = made["severity"].value_counts()
    print(f"  critical_score={cs:<5}  critical {int(counts.get('critical', 0)):>3}   "
          f"warning {int(counts.get('warning', 0)):>3}   info {int(counts.get('info', 0)):>3}")

severity as shipped (critical_score=2.0)
severity
critical    184

role weights in wk.detect.DEFAULT_ROLE_WEIGHT
  wan-primary      1.0
  server-uplink    0.9
  wan-secondary    0.6
  office-vlan      0.5
  backup-storage   0.3

severity split as critical_score moves
  critical_score=2.0    critical 184   warning   0   info   0


  critical_score=8.0    critical  44   warning 118   info  22
  critical_score=12.0   critical  30   warning  97   info  57


  critical_score=15.0   critical  28   warning  90   info  66


184 則，全部 critical。原因不在資料。`AlertPolicy.critical_score` 預設 2.0，而任何能通過門檻
4.0 的告警，peak score 至少是 4，`min(4/2, 2)` 已經頂到上限 2，第一項就貢獻 `0.5 x 2 = 1.0`，
直接達到 critical 的門檻。duration 和 role weight 完全沒有機會參與判斷。

一個把每則都標成 critical 的嚴重度模型，和沒有嚴重度模型是同一件事，而且更糟，因為它看起來像有在
分級。修法是把 `critical_score` 拉到明顯高於告警門檻的位置，例如 12 到 15，讓 peak score 4 到 6
的告警落在中段，duration 和 role 才有作用。上面那三列就是把旋鈕轉過去之後的分布，改完重跑，數字
對不上就是你改到別的地方去了。

那張 role weight 表是工作坊的起點，不是事實。它應該由營運方擁有，而且每次網路架構變動都要重新看
一次。

### 為什麼 role weight 必須存在

同一個 score 4.5，出現在 backup-storage 的凌晨兩點，和出現在 wan-primary 的上午十點，不是同一件
事。前者多半是備份、快照或索引重建，影響面是一台儲存設備和一個沒有人在等的工作；後者是對外連線，
影響面是所有正在用系統的人。

用預設權重算，同樣 peak 4.5、同樣持續 30 分鐘：backup-storage 拿到
`0.5x2 + 0.3x1 + 0.4x0.3 = 1.42`，wan-primary 拿到 `0.5x2 + 0.3x1 + 0.4x1.0 = 1.70`。兩者都是
critical，因為前面說的缺陷還在。把 `critical_score` 修好之後再算一次，差距才會落到不同的分級上。

嚴重度是把技術偏離翻譯成業務語言的那一步。翻譯錯了，收到的人會用「這條規則不準」來處理，然後整條
規則就死了。

## 10. 一則通知該帶什麼

早上的講義列過十項：metric 與當前值、baseline、門檻、分數、開始時間、持續時間、受影響的 instance、
dashboard 連結、runbook、規則版本。缺一項，收到的人就得自己把偵測器早就握有的上下文重建一次，
而那正是凌晨三點最不該做的事。

下一格印 `wk.notification_text()` 的輸出，挑的是持續最久的那一則真實告警，事件 F 打在 wan-primary
上。這裡挑的不是分數最高的那一則：整個月分數最高的告警是一則 false alert，這件事本身值得記住，
peak score 不是可信度。

把印出來的內容對著上面那十項逐項核對，缺的記下來。

In [12]:
matched = al[al["matched_event"]]
pick = matched.loc[matched["duration_s"].idxmax()]

print(wk.notification_text(
    pick,
    metric="score_max over traffic_bps + shape features",
    baseline_desc=f"1h rolling median/MAD, threshold {THRESHOLD}, {N_CONSEC} consecutive samples",
    dashboard_url=f"http://localhost:3000/d/aiops-lab02?var-port={pick['port_id']}",
    runbook="runbooks/traffic-deviation.md",
))
print()
print("evidence carried on the alert record, per feature:")
print(pick[[c for c in matched.columns if c.startswith("evidence_")]].round(2).to_string())

[CRITICAL] score_max over traffic_bps + shape features on port-id7427 (wan-primary)
  detector      multi_feature  vs  1h rolling median/MAD, threshold 4.0, 2 consecutive samples
  deviation     21.30 (peak at 2026-02-17 17:30:00)
  started       2026-02-17 16:05:00
  duration      105 min over 21 samples
  severity      critical (score 2.0)
  dashboard     http://localhost:3000/d/aiops-lab02?var-port=port-id7427
  runbook       runbooks/traffic-deviation.md

evidence carried on the alert record, per feature:
evidence_score_robust        21.3
evidence_score_seasonal     21.15
evidence_score_broadcast     2.28
evidence_score_multicast     1.38
evidence_score_pktsize      17.77


有的：metric 名稱、門檻、分數、開始時間、持續時間、instance 與它的角色、dashboard、runbook。
**缺三項**：metric 的當前值（只有分數，沒有 traffic_bps 實際是多少）、baseline 的實際數值
（只有 baseline 的文字描述，沒有那一刻的 `robust_center`）、規則版本。

缺前兩項的後果很具體。收到通知的人沒辦法判斷這是 200 MB/s 變成 400 MB/s，還是 2 MB/s 變成
4 MB/s，兩者的 z-score 可以一樣而處理方式完全不同。缺規則版本的後果要一個月後才會出現：有人調過
門檻，事後翻通知記錄卻不知道當時跑的是哪一版。

還有一個藏在細節裡的問題。`threshold 4.0` 這幾個字之所以會出現，只是因為呼叫端把它寫進了
`baseline_desc` 這個字串。上一格用 f-string 內插 `THRESHOLD` 就是為了避免有人改了門檻卻忘了改
字串，讓通知說謊。凡是要人手動同步的欄位，遲早會不同步。

### 補回去，在 Grafana 上

診斷完就要修，不然這一節只是抱怨。下面五步把當前值和規則版本補進去：

1. 左側 nav > Alerting > Notification policies，看 Default policy 那一列送到哪一個 contact
   point，記下名字。
2. 切到 Contact points，找同名的那一個按 Edit，展開 Optional Email settings，找到 `Message`
   欄位，它預設是空的，空的時候 Grafana 用內建模板。
3. 貼上下面這段，存到 `Message` 裡：

   ```text
   {{ range .Alerts }}
   {{ .Annotations.summary }}
     instance   {{ .Labels.port_id }} ({{ .Labels.port_role }})
     detail     {{ .Annotations.description }}
     version    {{ .Annotations.rule_version }}
     started    {{ .StartsAt }}
   {{ end }}
   ```

4. 按 Save contact point。想確認模板長什麼樣，用同一頁的 Test 送一則測試通知，不必等真的有規則
   Firing。
5. 回 Alerting > Alert rules，打開你在第 7 節建的那條規則按 Edit。在寫 annotation 的那一區，把
   description 填成 `Port {{ $labels.port_id }} 現在是 {{ $values.B }}，對照 dashboard 上的
   robust_center`，其中 `$values.B` 就是規則 B 段 reduce 出來的當前值；再按 Add custom
   annotation，name 填 `rule_version`，value 填今天日期加你的名字。Save rule and exit。

兩種語法不要弄混：規則自己的 annotation 用 `{{ $values.B }}` 和 `{{ $labels.port_id }}`，
contact point 的模板用 `{{ .Annotations.x }}` 和 `{{ .Labels.x }}`。前者在規則求值的時候展開，
後者在送通知的時候展開，寫錯的地方不會報錯，只會在通知裡留下一段空白。

## 11. 人的檢查點：門檻是商業決策

這一節結束在四個問題上。分成小組，每組挑兩個，五分鐘後各講一句結論。

**一、交換率。** 用 sweep 表指出你要站的那一格。從門檻 4.0 走到 3.0 的那筆帳第 2 節算過：多抓一個
事件，一個月多付 329 則告警，每天多 11 則。你的團隊願意用多少則告警換一個事件？講出一個數字，
然後說明這個數字是怎麼從你的人力、輪班和告警處理時間推出來的。順帶回答：既然那個事件用一條零誤報
的固定規則就抓得到，你為什麼還要考慮動門檻？

**二、誰有權決定。** 門檻現在寫在偵測器裡，由寫程式的人設定。它應該由誰決定，寫程式的人、值班的
人，還是服務的擁有者。這個決定記錄在哪裡，誰簽字，下次有人想改的時候要問誰。答案是「沒有人，就是
那樣設的」的團隊，今天要帶回去的第一個待辦事項就在這裡。

**三、不對稱的代價。** 漏掉一次 link_quality_issue 的代價，和一個月多收 329 則告警的代價，哪一個
大？你的答案在 wan-primary 上和在 backup-storage 上一樣嗎？如果不一樣，你的偵測器現在有沒有辦法
表達這個差別？

**四、退場條件。** 什麼證據會讓你把一條規則關掉？沒有退場條件的規則只會累積，三年之後沒有人記得
為什麼有那條規則，也沒有人敢刪。現在替你今天新建的那條規則寫一句退場條件，例如「連續兩個月誤報率
超過九成就下線重審」。

## 12. 沒有真值的那一天

前面每一節用到的數字，event recall、detection delay、precision、sweep 表、上一節那場辯論，全部是
拿告警去對 `synthetic_event_catalog.csv` 算出來的。星期一早上你不會有那個檔案。沒有人替你標好
十個事件的起訖時間，也沒有人告訴你哪一則告警是真的。

下一格做一個誠實的模擬。先把真值窗砍掉，只留下當初真的會進 ticket 系統的那幾段，拿同一批 182 則
告警重算一次記分卡；再把 182 則按 `peak_score` 排序取前 20 則，那是沒有真值也做得到的複核。

In [13]:
TICKETED = ["D", "F", "G"]            # 使用者抱怨過，一定進了 ticket 系統
TICKETED_PLUS = ["D", "E", "F", "G"]  # 再加上 NOC 自己在線路品質報表上查出來的那一件

keys = ["events_total", "events_detected", "event_recall", "event_precision",
        "false_alerts_unexplained", "point_recall"]
rows = []
for name, ids in [("full catalogue", None), ("ticketed only", TICKETED),
                  ("ticketed + NOC", TICKETED_PLUS)]:
    sub = incidents if ids is None else incidents[incidents["event_id"].isin(ids)]
    got, _, _ = wk.score(alerts, sub, SPAN, labels=flags["label"],
                         truth=wk.truth_mask(allp, sub), changes=calendar)
    rows.append({"catalogue": name, "windows": len(sub),
                 **{k: round(got[k], 4) if isinstance(got[k], float) else got[k] for k in keys}})

print("the same 182 alerts, scored against three different ground truths")
print(pd.DataFrame(rows).to_string(index=False))

cols = (["port_id", "port_role", "fire_time", "duration_s", "peak_score", "matched_event"]
        + [c for c in al.columns if c.startswith("evidence_")])
top = al.sort_values("peak_score", ascending=False).head(20)[cols].copy()
top["duration_min"] = (top.pop("duration_s") / 60).astype(int)
numeric = top.select_dtypes("number").columns
top[numeric] = top[numeric].round(1)
print()
print("top 20 alerts by peak score, the review you can run without a catalogue")
print(top.to_string(index=False))
print()
print(f"of the top 20, {int(top['matched_event'].sum())} overlap a labelled incident window")

the same 182 alerts, scored against three different ground truths
     catalogue  windows  events_total  events_detected  event_recall  event_precision  false_alerts_unexplained  point_recall
full catalogue       16            16               15        0.9375           0.0879                       166        0.4204
 ticketed only        7             7                7        1.0000           0.0385                       175        0.5435
ticketed + NOC        8             8                7        0.8750           0.0385                       175        0.5000

top 20 alerts by peak score, the review you can run without a catalogue
    port_id      port_role           fire_time  peak_score  matched_event  evidence_score_robust  evidence_score_seasonal  evidence_score_broadcast  evidence_score_multicast  evidence_score_pktsize  duration_min
port-id7427    wan-primary 2026-02-01 00:20:00        50.0          False                    0.6                      0.5                      50

看第二列。只憑使用者抱怨過的三個事件重建真值，recall 讀出來是滿分，而真實的 recall 是 0.9375。
差在 E link_quality_issue，那是一個沒有人打電話進來、只有錯誤率報表看得出來的故障，也正好是這個
偵測器唯一抓不到的那一個。**你偵測不到的事件，通常也是最不容易進 ticket 系統的事件**，這兩件事
同源，所以弱標籤下的 recall 是系統性高估，不是隨機誤差。第三列把 E 補進來，recall 就掉到 0.8750，
比真值還低，因為分母小、一個事件就佔八分之一。precision 則兩列都掉到 0.0385，方向和 recall 相反：
真值窗越少，越多告警被判成誤報。

沒有真值的時候能做的事印在第二張表上。這是早上講的 top-N review，也是唯一一種不需要標籤就能執行的
評估。表上的 `evidence_` 欄位是每個特徵在那段區間裡的最大絕對分數，判讀就是看這幾欄：
`evidence_score_broadcast` 獨大的多半是 L2 問題，`evidence_score_robust` 和
`evidence_score_seasonal` 同時很大的是真的流量搬家，`evidence_score_pktsize` 獨大的是封包形狀變了
而總量沒變。

現在動手。蓋住 `matched_event` 那一欄，從上往下判十則，每一則寫一個字：真、假、或不確定。判完再
掀開答案對。有兩處值得先知道會踩到什麼。前 20 則裡有 14 則落在真值窗內，剩下 6 則裡有 5 則同時
發生在 2026-02-22 21:40，五個 port 一起，而事件 H 的真值窗在 21:30 結束。那五則在記分卡上是誤報，
在工程上是同一場 flooding 的尾巴，差別只在標記窗的邊界畫在哪裡。另外那一則排在整張表的第一列，
2026-02-01 00:20 打在 port-id7427 上，broadcast 與 multicast 同時頂到 50 的上限，而沒有任何真值窗
蓋住它。它是真的異常，還是夜間 degenerate scale 的產物，留給你判。這就是星期一早上你手上會有的
東西：一張排序過的表，幾欄證據，和你自己的判斷。

## 推上線

這裡送出的是 Lab 01 與 Lab 02 兩張 dashboard 需要的欄位聯集，加上四個 scalar，讓 aiops-lab02 上
的四塊 stat tile 讀得到記分卡。`packets_pps` 也在裡面，雖然沒有任何 panel 查它，它是 Lab 01
publish 的欄位之一，聯集就該有它。

四塊 tile 都設了 percentunit 或對應的 unit，所以 `event_recall` 0.9375 在畫面上是 93.75%。

In [14]:
COLUMNS = [
    "traffic_bps", "packets_pps", "error_rate", "discard_rate", "errors_pps",
    "avg_pkt_bytes", "broadcast_ratio", "multicast_ratio",
    "roll_center", "roll_lo", "roll_hi",
    "robust_center", "robust_lo", "robust_hi",
    "seasonal_center", "seasonal_lo", "seasonal_hi",
    "peer_center", "peer_lo", "peer_hi",
    "score_rolling", "score_robust", "score_seasonal", "score_peer",
    "score_broadcast", "score_multicast", "score_pktsize",
    "score_cusum", "cusum_signal",
    "score_max", "label", "alert", "is_incident", "is_planned",
]

wk.publish(allp, COLUMNS, "lab02_detection", scalars={
    "eval_event_recall": summary["event_recall"],
    "eval_alerts_per_day": summary["alerts_per_day"],
    "eval_mttd_min": summary["median_detection_delay_min"],
    "eval_false_unexplained": summary["false_alerts_unexplained"],
})

print()
print("Lab 02  http://localhost:3000/d/aiops-lab02")
print("Lab 01  http://localhost:3000/d/aiops-lab01   (same publish keeps it populated)")

published 43,200 rows as 'lab02_detection'
  archive   outputs/workshop/lab02_detection.csv
  drop zone outputs/prometheus-dropzone/current_results.csv
  columns   traffic_bps, packets_pps, error_rate, discard_rate, errors_pps, avg_pkt_bytes, broadcast_ratio, multicast_ratio, roll_center, roll_lo, roll_hi, robust_center, robust_lo, robust_hi, seasonal_center, seasonal_lo, seasonal_hi, peer_center, peer_lo, peer_hi, score_rolling, score_robust, score_seasonal, score_peer, score_broadcast, score_multicast, score_pktsize, score_cusum, cusum_signal, score_max, label, alert, is_incident, is_planned, eval_event_recall, eval_alerts_per_day, eval_mttd_min, eval_false_unexplained
  query     aiops_python_result{column="traffic_bps"}

Lab 02  http://localhost:3000/d/aiops-lab02
Lab 01  http://localhost:3000/d/aiops-lab01   (same publish keeps it populated)


## 把真值窗畫成 annotation

`Ground truth - known incident windows` 那一排是 state timeline，只在你選定某個 port 的時候有
意義，換一張 dashboard 就沒了。Grafana 的 annotation 是另一種掛法：它掛在 dashboard 上，每一張
panel 都會出現同一塊灰底區間，滑鼠移上去看得到事件名稱。

sim time 要先換算成牆鐘時間才貼得準，`wk.event_annotations` 吃一個 `ReplayClock` 就是為了這件事。
`to_wall` 算的是第一輪的位置，而 exporter 會一直繞圈，所以下面多做一步：把整批往後推整數輪，讓它
落在你現在正在看的這一輪上。過了一輪之後重跑這一格就好。

In [15]:
import time

clock = wk.grafana.ReplayClock.load()
client = wk.GrafanaClient()

if clock is None:
    print("no replay_state.json yet: start the exporter, then re-run this cell")
elif not client.healthy():
    print("Grafana is not reachable on localhost:3000, skipping the annotation push")
else:
    items = wk.event_annotations(incidents, tag="aiops-incident", clock=clock)
    pass_ms = int(clock.span_s() / clock.speed_x * 1000)
    shift = int((time.time() - clock.wall_start) * 1000) // pass_ms * pass_ms
    for item in items:
        item["time"] += shift
        item["timeEnd"] += shift
        item["dashboardUID"] = "aiops-lab02"

    pushed = client.annotate(items, clear_tag="aiops-incident")
    print(f"pushed {pushed} incident regions onto aiops-lab02")
    print(f"one replay pass is {clock.pass_duration_min():.0f} wall minutes; "
          f"re-run this cell after that and the regions move with it")

pushed 16 incident regions onto aiops-lab02
one replay pass is 60 wall minutes; re-run this cell after that and the regions move with it


## 收尾

離開之前把兩件事還原。第 6 節暫時改動的門檻和 pending period 要改回 `4` 與 `10m`，第 8 節如果動過
time interval 要改回 00:30 到 05:00。你自己新建的第四條規則留著，那是你今天唯一從零寫出來的東西。

drop zone 現在裝的是 Lab 02 的結果，Lab 01 需要的欄位也在裡面，所以那兩張 dashboard 都有資料。
Lab 00 的 toy signal 不在這一份裡，它的 panel 會是空的，那是正常的。下一次 publish 會整份覆蓋掉。